llmfoundry/models/mpt/modeling_mpt.py

```diff
@@ -448,6 +454,142 @@ def construct_blocks(self, config: MPTConfig) -> nn.ModuleList:
+            )
+
+        for i in range(config.n_layers):
+            module_name = model_modules_order_expanded[i]
+            override_config = {}
+            if module_name != 'default':
+                override_config = copy.deepcopy(
+                    config.block_overrides['overrides'][module_name],
+                )
+                override_attn_config = override_config.get('attn_config', None)
+                if override_attn_config is not None and 'reuse_kv_layer_idx' in override_attn_config:
+                    if override_attn_config['reuse_kv_layer_idx'] >= 0:
+                        raise ValueError(
+                            f'The relative index of kv layer to reuse, {override_attn_config["reuse_kv_layer_idx"]=}, should be negative.',
+                        )
+                    reuse_kv_layer_idx = i + override_attn_config[
+                        'reuse_kv_layer_idx']
+                    if reuse_kv_layer_idx < 0:
+                        raise ValueError(
+                            f'The absolute index of kv layer to reuse, {reuse_kv_layer_idx} should be non-negative.',
+                        )
+                    override_attn_config['reuse_kv_layer_idx'
+                                        ] = reuse_kv_layer_idx
+                    if self.kv_cache_layers is None:
+                        self.kv_cache_layers = set()
+                    self.kv_cache_layers.add(reuse_kv_layer_idx)
+            layer_description_list.append([
+                i,
+                module_name,
+                self._get_overrides_for_logging(override_config),
+            ],)
+            new_block_args = self._override_block_args(
+                block_args,
+                override_config,
+                config.allowed_block_overrides,
+            )
+            module_list.append(
+                MPTBlock(
+                    device=config.init_device,
+                    **new_block_args,
+                ),
+            )
+        log.info(
+            'The following is a summary of overrides per layer.\n' + tabulate(
+                layer_description_list,
+                headers=['idx', 'name', 'overrides'],
+            ),
+        )
+        return nn.ModuleList(module_list)
+
+    def _get_overrides_for_logging(
+        self,
+        override_config: Dict[str, Any],
+    ) -> List[dict[str, str]]:
+        overrides_list = []
+        for k, v in override_config.items():
+            if isinstance(v, dict):
+                overrides_list.extend(self._get_overrides_for_logging(v))
+            else:
+                overrides_list.append({k: v})
+        return overrides_list
+
+    def _override_block_args(
+        self,
+        block_args: Dict[str, Any],
+        override_config: Dict[str, Any],
+        allowed_block_overrides: set,
+    ) -> Dict[str, Any]:
+        new_block_args = override_config | block_args
```

Comment:
would it be simpler to make a shallow copy of `block_args` here and then just iterate over the keys in `override_config` (keeping the recursive bit)?

Reply:
Discussed offline, added test.